contracts/Allowlist.sol

```diff
@@ -100,69 +184,189 @@ contract Allowlist {
       conditionIdx++
     ) {
       Condition memory condition = _conditions[conditionIdx];
-      addCondition(condition);
+      addConditionWithoutValidation(condition);
     }
   }
 
-  function deleteCondition(uint256 conditionIdx) public onlyOwner {
-    Condition memory lastCondition = conditions[conditions.length - 1];
-    conditions[conditionIdx] = lastCondition;
-    conditions.pop();
+  /**
+   * @notice Delete a condition given an ID
+   * @param conditionId The ID of the condition to delete
+   */
+  function deleteCondition(string memory conditionId) public onlyOwner {
+    string memory lastConditionId = conditionsIds[conditionsIds.length - 1];
+    for (
+      uint256 conditionIdx;
+      conditionIdx < conditionsIds.length;
+      conditionIdx++
+    ) {
+      string memory currentConditionId = conditionsIds[conditionIdx];
+      if (Strings.stringsEqual(currentConditionId, conditionId)) {
+        conditionsIds[conditionIdx] = lastConditionId;
+        conditionsIds.pop();
+        delete conditionById[conditionId];
+        return;
+      }
+    }
+    revert("Cannot find condition with that ID");
   }
 
-  function deleteConditions(uint256[] memory conditionsIdxes) public onlyOwner {
-    // Reverse sort condition indexes
-    uint256[] memory conditionIndexesDescending = Quicksort.sort(
-      conditionsIdxes,
-      true
-    );
+  /**
+   * @notice Delete multiple conditions given a list of IDs
+   * @param _conditionsIds A list of condition IDs to delete
+   */
+  function deleteConditions(string[] memory _conditionsIds) public onlyOwner {
     for (
       uint256 conditionIdx;
-      conditionIdx < conditionIndexesDescending.length;
+      conditionIdx < _conditionsIds.length;
       conditionIdx++
     ) {
-      deleteCondition(conditionIndexesDescending[conditionIdx]);
+      string memory conditionId = _conditionsIds[conditionIdx];
+      deleteCondition(conditionId);
     }
   }
 
+  /**
+   * @notice Delete every condition
+   */
   function deleteAllConditions() public onlyOwner {
-    uint256 _conditionsLength = conditions.length;
+    uint256 _conditionsLength = conditionsIds.length;
     for (
       uint256 conditionIdx;
       conditionIdx < _conditionsLength;
       conditionIdx++
     ) {
-      conditions.pop();
+      string memory conditionId = conditionsIds[0];
+      deleteCondition(conditionId);
     }
   }
 
-  function updateCondition(uint256 conditionIdx, Condition memory condition)
-    public
-    onlyOwner
-  {
-    deleteCondition(conditionIdx);
+  /**
+   * @notice Update an existing condition
+   * @dev Reads condition.id to determine which condition to update
+   * @param condition The new condition
+   */
+  function updateCondition(Condition memory condition) public onlyOwner {
+    bool conditionExists = !Strings.stringsEqual(
+      conditionById[condition.id].id,
+      ""
+    );
```

Comment:
Maybe we could add a `conditionExists` function since we use this is a couple of places? It's a bit difficult to reason about these lines so might be good to have them in only one place

Reply:
good idea

Reply:
Implemented